### PR TITLE
Add mutex to KernelMap Cache

### PR DESCRIPTION
**Pull request by chaeyeunpark**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** KernelMap class provides the best kernel to use based on [number of qubits], [threading], and memory alignment. To generate the kernel information quickly, it uses an internal cache mechanism. However, this violated const correctness and raised a multithreading bug. This PR fixes this bug.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane_lightning/src/simulator/KernelMap.hpp
Comment (mlxd) [inline pennylane_lightning/src/simulator/KernelMap.hpp:190]: Can we add a docstring here?

File: pennylane_lightning/src/simulator/KernelMap.hpp
Comment (mlxd) [inline pennylane_lightning/src/simulator/KernelMap.hpp:359]: I assume we are safe unlocking here for multiple threads accessing this function and `updateCache` simultaneously? I think I can reason it all looks good, but just confirming the lock at line 201 is good.

File: pennylane_lightning/src/tests/Test_KernelMap.cpp
Comment (mlxd) [inline pennylane_lightning/src/tests/Test_KernelMap.cpp:186]: All good with this number?

File: pennylane_lightning/src/tests/Test_KernelMap.cpp
Comment (mlxd) [inline pennylane_lightning/src/tests/Test_KernelMap.cpp:148]: Should parameterise the tests based on num_qubits or is it fine to handle it in the same case?

**Review by mlxd — COMMENTED** [review]
Thanks @chaeyeunpark 
Looks good, but just a few queries before approving.

File: pennylane_lightning/src/simulator/KernelMap.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/simulator/KernelMap.hpp:359]: Yeap. I unlocked the mutex here to ensure that an element of cache is constructed outside of mutex. We lock the cache again when we insert this element.

File: pennylane_lightning/src/tests/Test_KernelMap.cpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/tests/Test_KernelMap.cpp:148]: I don't think we need to do so. I just randomly generated an extreme case that calls `getKernelMap` lots of times simultaneously from multiple threads. This test case already raises an error for the previous code.

File: pennylane_lightning/src/tests/Test_KernelMap.cpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/tests/Test_KernelMap.cpp:186]: I just made this as a `constexpr` variable instead of hard coding here. Runtime-wise, it is fine (less than 0.1 sec).

File: pennylane_lightning/src/simulator/KernelMap.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/simulator/KernelMap.hpp:197]: Added.

**Review by AmintorDusko — APPROVED** [review]
Thank you, @chaeyeunpark.

File: pennylane_lightning/src/tests/Test_KernelMap.cpp
Comment (mlxd) [reply pennylane_lightning/src/tests/Test_KernelMap.cpp]: Sounds good -- thanks!

**Review by mlxd — APPROVED** [review]
Nice work @chaeyeunpark

**Comment by AmintorDusko** [comment]
Please wait for PR #383 to be merged before merging this PR.

**Comment by chaeyeunpark** [comment]
Thanks all for cool reviews!

**Pull request merged**
